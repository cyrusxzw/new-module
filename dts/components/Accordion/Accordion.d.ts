import React from 'react';
import type { AccordionProps } from './Accordion.types';
declare const Accordion: React.ForwardRefExoticComponent<
  AccordionProps & React.RefAttributes<HTMLDivElement>
>;
export { Accordion };
