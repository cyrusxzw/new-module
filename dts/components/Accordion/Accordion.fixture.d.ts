/// <reference types="react" />
export declare const AccordionFixture: {
  id: string;
  items: (
    | {
        id: string;
        heading: string;
        content: string;
      }
    | {
        id: string;
        heading: string;
        content: JSX.Element;
      }
  )[];
};
