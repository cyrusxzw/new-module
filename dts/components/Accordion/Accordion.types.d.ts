import { ReactNode } from 'react';
import type { Themes } from '~/types';
declare type AccordionItem = {
  content: ReactNode;
  heading: string;
  id: string;
};
declare type AccordionProps = {
  className?: string;
  id?: string;
  items?: AccordionItem[];
  theme?: Themes;
  wrapperClass?: string;
};
export type { AccordionProps };
