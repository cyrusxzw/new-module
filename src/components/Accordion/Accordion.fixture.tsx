import React from 'react';

export const AccordionFixture = {
  id: 'testing-id',
  items: [
    {
      id: '1',
      heading: 'What are the granting programs available for application?',
      content:
        'The Aesop Foundation will consider support for organisations working in Australia. While the majority of our grants are awarded to organisations that are invited to apply, we remain open to hearing new ideas from potential partners. If you believe your organisationâ€™s project meets the objectives and guidelines of the Foundation we warmly invite you to submit a letter of enquiry.',
    },
    {
      id: '2',
      heading: 'What organisations are considered for applications?',
      content:
        'The Aesop Foundation supports its partners through two granting programs. The Companion Grants program provides up to AUD30,000, usually for a single year. The Partnerships Grants program provides between AUD100,000-300,000 per year for up to five years.',
    },
    {
      id: '3',
      heading: 'What are the guidelines for an application?',
      content: (
        <React.Fragment>
          <p>
            We respectfully request that you review the following guidelines to
            ensure close alignment prior to submitting a letter of enquiry:
          </p>
          <p>
            Applicant organisations must be a DGR Item 1 entity, operating in
            Australia and a charity registered with the ACNC.
          </p>
          <p>
            The project must meet, or make a contribution to meeting, one or
            both of the objectives of the Foundation, as follows',
          </p>
        </React.Fragment>
      ),
    },
  ],
};
