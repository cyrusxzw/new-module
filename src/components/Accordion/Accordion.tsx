import React, { forwardRef } from 'react';
import cx from 'classnames';
import {
  Accordion as AccessibleAccordion,
  AccordionItem,
  AccordionItemHeading,
  AccordionItemButton,
  AccordionItemPanel,
} from 'react-accessible-accordion';
import { Collapse } from 'react-collapse';
import { Icon } from '~/components/Icon';
import { isObjectPopulatedArray } from '~/utils/objects';
import type { AccordionProps } from './Accordion.types';
import styles from './Accordion.module.css';

const Accordion = forwardRef<HTMLDivElement, AccordionProps>(
  ({ className, id, items, theme = 'dark', wrapperClass }, ref) => {
    const [activeNodes, setIsActiveNodes] = React.useState([]);

    if (!isObjectPopulatedArray(items)) {
      return null;
    }

    const classSet = cx(styles.base, styles[theme], className);
    /** handleOnChange gets passed to the AccessibleAccordion dependency */
    /* istanbul ignore next */
    const handleOnChange = nodes => setIsActiveNodes(nodes);
    const checkIsActive = (id: string) => activeNodes.includes(id);

    return (
      <div
        className={cx(wrapperClass)}
        data-testid="data-testid-Accordion"
        id={id}
        ref={ref}
      >
        <AccessibleAccordion
          allowMultipleExpanded={true}
          allowZeroExpanded={true}
          className={classSet}
          onChange={handleOnChange}
        >
          {items.map(({ content, heading, id }) => (
            <AccordionItem className={styles.item} key={id} uuid={id}>
              <AccordionItemHeading className={styles[theme]}>
                <AccordionItemButton className={styles.button}>
                  <span className={styles.headingCopy}>{heading}</span>
                  <Icon
                    className={styles.icon}
                    height={16}
                    name="downArrow"
                    theme={theme}
                    width={16}
                  />
                </AccordionItemButton>
              </AccordionItemHeading>
              <AccordionItemPanel className={styles.panel}>
                <Collapse
                  isOpened={checkIsActive(id)}
                  theme={{ collapse: styles.collapse }}
                >
                  <div className={cx(styles.content, styles[theme])}>
                    {content}
                  </div>
                </Collapse>
              </AccordionItemPanel>
            </AccordionItem>
          ))}
        </AccessibleAccordion>
      </div>
    );
  },
);

export { Accordion };
