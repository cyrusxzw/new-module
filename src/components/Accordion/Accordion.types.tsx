import { ReactNode } from 'react';
import type { Themes } from '~/types';

type AccordionItem = {
  content: ReactNode;
  heading: string;
  id: string;
};

type AccordionProps = {
  className?: string;
  id?: string;
  items?: AccordionItem[];
  theme?: Themes;
  wrapperClass?: string;
};

export type { AccordionProps };
