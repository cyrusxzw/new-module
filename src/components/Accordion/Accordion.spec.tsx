import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { AccordionFixture } from './Accordion.fixture';
import { Accordion } from './Accordion';

describe('<Accordion />', () => {
  it('should return `null` if the length of the `items` prop is 0', () => {
    render(<Accordion items={[]} />);

    const accordion = screen.queryByTestId(/data-testid-Accordion/i);

    expect(accordion).not.toBeInTheDocument();
  });

  describe(`DOM attributes`, () => {
    let accordionWrapper: HTMLElement = null;

    const className = 'class1';
    const id = 'my-accordion-id';
    const theme = 'dark';
    const wrapperClass = 'the-great-cornholio';

    beforeEach(() => {
      accordionWrapper = render(
        <Accordion
          className={className}
          id={id}
          items={AccordionFixture.items}
          theme={theme}
          wrapperClass={wrapperClass}
        />,
      // eslint-disable-next-line testing-library/no-node-access
      ).container.firstChild as HTMLElement; // this is to access the root level element of the component
    });

    it(`should assign the wrapperClass to the wrapping div`, () => {
      expect(accordionWrapper).toHaveClass(wrapperClass);
    });

    it(`should assign the id to the wrapping div`, () => {
      expect(accordionWrapper).toHaveAttribute('id', id);
    });

    it(`should correctly set the classes on the Accordion`, () => {
      // eslint-disable-next-line testing-library/no-node-access
      expect(accordionWrapper.firstChild).toHaveClass(theme, className);
    });
  });

  describe(`Headings`, () => {
    let headings: HTMLElement[] = null;

    beforeEach(() => {
      render(<Accordion items={AccordionFixture.items} />);

      headings = screen.getAllByRole('heading');
    });

    it(`should render the correct number of headings`, () => {
      expect(headings.length).toEqual(3);
    });

    it(`should render with the correct heading text`, () => {
      AccordionFixture.items.forEach(({ heading }, index) => {
        expect(headings[index]).toHaveTextContent(heading);
      });
    });
  });

  describe(`Content`, () => {
    let contentSections: HTMLElement[] = null;

    beforeEach(() => {
      render(<Accordion items={AccordionFixture.items} />);
      contentSections = screen.getAllByTestId('Accordion-content-section');
    });

    it(`should load with all the content hidden`, () => {
      contentSections.forEach(contentSection => {
        expect(contentSection).toBeInTheDocument();
        expect(contentSection).not.toBeVisible();
      });
    });

    it(`should expand/collapse a section when the heading is clicked`, () => {
      const sectionIndex = 0;
      const firstSectionButton = screen.getByRole('button', {
        name: new RegExp(AccordionFixture.items[sectionIndex].heading, 'i'),
      });
      const contentSection = screen.getAllByTestId('Accordion-content-section')[
        sectionIndex
      ];

      userEvent.click(firstSectionButton);
      expect(contentSection).toBeVisible();
      userEvent.click(firstSectionButton);
      expect(contentSection).not.toBeVisible();
    });
  });
});
