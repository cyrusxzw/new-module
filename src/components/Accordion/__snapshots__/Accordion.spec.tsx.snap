// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> should render base component correctly 1`] = `
<div>
  <div
    class=""
    data-testid="data-testid-Accordion"
    id="testing-id"
  />
</div>
`;
